lightning/src/offers/invoice_request.rs

```diff
@@ -1159,8 +1159,13 @@ impl TryFrom<PartialInvoiceRequestTlvStream> for InvoiceRequestContents {
 			return Err(Bolt12SemanticError::MissingAmount);
 		}
 
-		offer.check_quantity(quantity)?;
-		offer.check_amount_msats_for_quantity(amount, quantity)?;
+		match offer.amount() {
+			Some(Amount::Currency {..}) => return Err(Bolt12SemanticError::UnsupportedCurrency),
```

Comment:
Unfortunately, it looks like we can't fail like this because this `TryFrom` implementation is also used when parsing a `Bolt12Invoice`. So if we send an amount-less request for an `Offer` with a currency, we'll fail to parse the invoice that is sent back. Here's a [branch](https://github.com/jkczyz/rust-lightning/commits/2024-08-invoice-from-offer-with-currency/) with a test that demonstrates this. Feel free to cherrypick the commit.

I guess we need to move the these checks higher up in the parsing code, after the contents has been created. Haven't checked if that will have any unexpected behavior.

Reply:
@TheBlueMatt @tnull I wonder if we should allow parsing an `InvoiceRequest` with any amount and rather check the amount when handling it (i.e., in `ChannelManager`). Otherwise, once currency support is added, it would require looking up exchange rates at parsing time, which seems wrong. Also, if no amount is specified in the request, then that is where the issuer would convert from the offer amount. Thoughts?

Reply:
@jkczyz I moved the checks higher up in the parsing code as you suggested. But, after cherry-picking the commit and running the tests, the `parses_invoice_with_currency` test fails. Given the context, it looks like checking amounts during handling might be more fitting. If so, I can include this change in this PR.

Reply:
Sure, let's go ahead and do that. Longer term it makes more sense, IMO.